New package: ghostscript

darealshinji opened the pull request:
Since the upstream developers don't seem to care much about supporting MinGW, and especially not about cross-compiling, and since they use a kind of ugly build system, a lot of stuff needed to be patched.
The patch is based on the patches from the [MinGW package](https://github.com/Alexpux/MINGW-packages/tree/master/mingw-w64-ghostscript), plus it enables cross-compiling.
If this gets merged, [libspectre](https://github.com/darealshinji/mxe/tree/libspectre-new) could be added too, which is a wrapper library for ghostscript.

darealshinji commented:
Rebased commits into a single one.

tonytheodore commented on src/ghostscript.mk:
```diff
@@ -0,0 +1,74 @@
+
+    '$(TARGET)-gcc' \
+        -W -Wall -Werror -pedantic `$(TARGET)-pkg-config --cflags ghostscript` \
+        $(if $(BUILD_STATIC),-DGS_STATIC_LIB, ) \
```
Move `-DGS_STATIC_LIB` to `Cflags.private:` section of `ghostscript.pc` so it's available to downstream packages. 
Use single invocation of `$(TARGET)-pkg-config --cflags --libs ghostscript`

darealshinji replied:
Done.

tonytheodore commented:
Looking good, builds on all four targets.

tonytheodore commented on src/ghostscript.mk:
```diff
@@ -0,0 +1,73 @@
+    cd '$(1)' && rm -rf freetype jpeg lcms2 libpng openjpeg tiff zlib
+    cd '$(1)' && $(LIBTOOLIZE) --force --copy --install
+    cd '$(1)' && autoconf -f -i
+    cd '$(1)' && ./configure $(MXE_CONFIGURE_OPTS) \
```
put `$(MXE_CONFIGURE_OPTS)` on its own line

tonytheodore commented on src/ghostscript.mk:
```diff
@@ -0,0 +1,73 @@
+     > '$(PREFIX)/$(TARGET)/lib/pkgconfig/ghostscript.pc'
+
+    '$(TARGET)-gcc' \
+        -W -Wall -Werror -pedantic $(if $(BUILD_STATIC),-DGS_STATIC_LIB, ) \
```
Remove `$(if $(BUILD_STATIC),-DGS_STATIC_LIB, )`, it's in the *pc file

starius commented:
> echo 'Libs.private: -lm -lidn -liconv -lpaper -ltiff -lpng -ljpeg -lopenjp2 -llcms2 -lz -lwinspool';) \ 

Can you replace it with `Requires:`? MXE has most (or all) of the libraries from this list.

starius commented:
Can you format the patch with `patch-tool-mxe`, please?

IMHO, if a patch is based on another (external) patch, it is better to keep it as two "commits" in a patch file. When the original patch updates, it will be easy to "rebase" changes against new version of upstream patch. I have separated your changes from the original patch. My patch file includes patches by MINGW-packages followed by further changes found in your initial patch.
https://gist.github.com/7f14838b3487cf8f282f03520b92a0a8

darealshinji commented:
I've updated ghostscript.mk and added your patch file.


>> echo 'Libs.private: -lm -lidn -liconv -lpaper -ltiff -lpng -ljpeg -lopenjp2 -llcms2 -lz -lwinspool';) \

>Can you replace it with Requires:? MXE has most (or all) of the libraries from this list.

Static linking against libgs will fail if I remove the `Libs.private` line. MXE has all these libraries but not all of them have a pkg-config file.

starius commented:
> but not all of them have a pkg-config file

Which ones do not have pc-files?
Can you use `Requires` for those who have, please?

starius commented:
Why do you build with `-j 1` instead of `-j $(JOBS)`?

starius commented:
I changed your .mk file to build in a separate dir and use new macros `SOURCE_DIR`, `BUILD_DIR` and `TEST_FILE`: https://gist.github.com/c2d4a6346eae98bdcbb6e0bcfadeab8a

darealshinji commented:
> Which ones do not have pc-files?

libm, libiconv, libpaper, libopenjp2 and libwinspool.

> Can you use Requires for those who have, please?

Like this? -> `Requires: libidn libtiff-4 libpng jpeg lcms2 zlib`

> Why do you build with -j 1 instead of -j $(JOBS)?

The build system doesn't seem to support more than 1 jobs. It'll eventually break at some point during the build.

darealshinji commented:
Building with `-j $(JOBS)` works on all targets. Still the problem with `BUILD_DIR`.

starius commented:
> Like this? -> Requires: libidn libtiff-4 libpng jpeg lcms2 zlib

Yes.

Also add a comment with https://github.com/mxe/mxe/issues/1446 near libraries linked directly through Libs.private, please.

> The build system doesn't seem to support more than 1 jobs. It'll eventually break at some point during the build.

Please report this to upstream and add link to the bug report to comment.

> Still the problem with BUILD_DIR.

What problems?

darealshinji commented:
> Also add a comment with #1446 near libraries linked directly through Libs.private, please.

Is this correct?
``` sh
     echo 'Requires: libidn libtiff-4 libpng jpeg lcms2 zlib'; \
     echo '# https://github.com/mxe/mxe/issues/1446'; \
     echo 'Libs.private: -lm -liconv -lpaper -lopenjp2 -lwinspool';) \
```

> Please report this to upstream and add link to the bug report to comment.

I think this only happened to me on the previous version of ghostscript. This isn't a problem anymore.

> What problems?

Out-of-tree building isn't supported:
```
[build]     ghostscript              i686-w64-mingw32.static                                              

Failed to build package ghostscript for target i686-w64-mingw32.static!
------------------------------------------------------------
./obj/aux_/echogs -e .dev -w- -l-dev2 -b -s -l-include -l./obj/page -l-obj ./obj/jpeggray ./obj/gdevjpeg.o
./obj/aux_/echogs -e .dev -a-  ./obj/jpeggray -include ./obj/sdcte
make[2]: *** No rule to make target `obj/mag16.dev', needed by `obj/devdevs.tr'.  Stop.
make[2]: Leaving directory `/home/djcj/mxe/tmp-ghostscript-i686-w64-mingw32.static/ghostscript-9.19.build_'
make[1]: *** [build-only-ghostscript_i686-w64-mingw32.static] Error 2
make[1]: Leaving directory `/home/djcj/mxe'
real	0m34.271s
user	0m29.100s
sys	0m1.688s
------------------------------------------------------------
[log]      /home/djcj/mxe/log/ghostscript_i686-w64-mingw32.static

make: *** [/home/djcj/mxe/usr/i686-w64-mingw32.static/installed/ghostscript] Error 1
```

But it works if I replace the `BUILD_DIR` macros with `SOURCE_DIR`.

starius commented:
> Shouldn't it be Requires.private?

IMHO, `*.private` is for dependencies not exposed to API. So the answer depends on whether these dependencies are exposed to API of ghostscript.

> Out-of-tree building isn't supported:

Please report this to upstream. I think, we can use in-tree building and provide a link to the report.

darealshinji commented:
I've tested a native Linux build on the unpatched source. The out-of-tree problem only occurs when the source was configured with `--disable-contrib`.

starius commented:
I am digging into this issue.

Other issues. I have found the following messages in the log:

```
configure: WARNING: unrecognized options: --enable-static, --disable-shared, --disable-doxygen
```

It doesn't understand static/shared switches.

```
configure: WARNING: Using "native" font scaler which is now deprecated (rather than freetype),
configure: WARNING: Support for this will be removed in a future release
```

Does it mean that it doesn't use freetype? Please, make sure that freetype provided by MXE is used.

starius commented:
The problem with out-of-tree build was fixed in https://github.com/LuaAndC/mxe/commit/7e218ad3b6240bb3cb2e4689c2c4232169cf0ad9 and https://github.com/LuaAndC/mxe/commit/508809b110bc9b9f7e537aad5afad64aadbb0230 .

It was caused by too precise grep pattern in configure.ac:

```shell
noncontribmakefiles=`find $srcdir -name '*.mak' -print | grep -v '^\./contrib/'`
```

Please `cherry-pick` or `am` my commits and rebase your branch against mxe:master.

darealshinji commented:
Changes were applied and branch was rebased.

starius commented:
List of installed files:

  * i686-w64-mingw32.shared/bin/libgs-9.dll
  * i686-w64-mingw32.shared/bin/test-ghostscript.exe
  * i686-w64-mingw32.shared/installed/ghostscript
  * i686-w64-mingw32.shared/lib/libgs.dll.a
  * i686-w64-mingw32.shared/lib/pkgconfig/ghostscript.pc
  * i686-w64-mingw32.static/bin/test-ghostscript.exe
  * i686-w64-mingw32.static/installed/ghostscript
  * i686-w64-mingw32.static/lib/libgs.a
  * i686-w64-mingw32.static/lib/pkgconfig/ghostscript.pc
  * x86_64-w64-mingw32.shared/bin/libgs-9.dll
  * x86_64-w64-mingw32.shared/bin/test-ghostscript.exe
  * x86_64-w64-mingw32.shared/installed/ghostscript
  * x86_64-w64-mingw32.shared/lib/libgs.dll.a
  * x86_64-w64-mingw32.shared/lib/pkgconfig/ghostscript.pc
  * x86_64-w64-mingw32.static/bin/test-ghostscript.exe
  * x86_64-w64-mingw32.static/installed/ghostscript
  * x86_64-w64-mingw32.static/lib/libgs.a
  * x86_64-w64-mingw32.static/lib/pkgconfig/ghostscript.pc

Looks good to me.

starius commented:
Thanks!
